agent/tool-scripts/postprocess/pidstat-convert

```diff
@@ -0,0 +1,118 @@
+#   - if the values repeat in subsequent sample, omit the value
+
+use strict;
+use warnings;
+
+# Check for an alternate tools library path for testing
+my $_test_alt_tools_lib;
+BEGIN {
+        my $_pbench_tspp_dir = $ENV{'pbench_tspp_dir'};
+	$_test_alt_tools_lib=$ENV{_TEST_ALTERNATE_TOOLS_LIBRARY};
+	if (not defined $_test_alt_tools_lib or not -d $_test_alt_tools_lib) {
+                $_test_alt_tools_lib = "$_pbench_tspp_dir";
+	}
+}
+use lib "$_test_alt_tools_lib";
+no lib ".";
+use File::Path qw(rmtree make_path);
+use File::Basename;
+use SysStat qw(get_pidstat_attributes);
+
+my %previous_stats; # used to detect repeated value
+my %current_stats;
+my @attributes = get_pidstat_attributes();
+my $tool_output_dir = shift; # the one and only argument to this script, the directory where these files should be written
+if (not defined $tool_output_dir) {
+	print "Directory to store PIDs is required\n";
+	exit 1;
+}
+my %file_handles;
+if ( -e $tool_output_dir . '/pids') {
+	rmtree($tool_output_dir . '/pids');
+}
+mkdir($tool_output_dir . '/pids');
+# read the STDIN which is expected to be STDOUT from 'pidstat -l -H -w -u -h -d -r -p ALL'
+my $prev_pid;
+while (my $line = <>) {
+	# example STDIN
+	#
+	###v11 no threads
+        ## Time        UID       PID    %usr %system  %guest   %CPU   CPU  minflt/s  majflt/s     VSZ     RSS   %MEM   kB_rd/s   kB_wr/s kB_ccwr/s iodelay   cswch/s nvcswch/s  Command
+        # 1547780606     0      2368    0.00    0.00    0.00    0.00     4      0.00      0.00  112808    4300   0.00      0.00      0.00      0.00       0      0.00      0.00  /usr/sbin/sshd -D
+        # 1547780606     0      2373    0.00    0.00    0.00    0.00     0      0.00      0.00  228772    7940   0.00      0.00      0.00      0.00       0      0.00      0.00  /usr/sbin/rsyslogd -n
+        # 1547780606     0      2380    0.00    0.00    0.00    0.00     7      0.00      0.00 1221772   15360   0.01      0.00      0.00      0.00       0      0.00      0.00  /usr/sbin/libvirtd
+        
+	###v11 with threads
+        # Time   UID      TGID       TID    %usr %system  %guest    %CPU   CPU  minflt/s  majflt/s     VSZ     RSS   %MEM   kB_rd/s   kB_wr/s kB_ccwr/s iodelay   cswch/s nvcswch/s  Command
+        # 1547852254     0         1         0    0.00    0.00    0.00    0.00     4      0.13      0.00  194520    7572   0.00      1.16     28.14      1.62      89      0.62      0.01  /usr/lib/systemd/systemd --switched-root --system --deserialize 22
+        # 1547852254     0         0         1    0.00    0.00    0.00    0.00     4      0.13      0.00  194520    7572   0.00      0.02      0.00      0.00      89      0.62      0.01  |__/usr/lib/systemd/systemd --switched-root --system --deserialize 22
+        #
+        if ( $line =~ /\s{0,1}(\d+)\s+\d+\s+(\S+|\S+\s+\S+)\s+(\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\s+\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\s+\d+\.\d+\s+\d+\.\d+\s+)(.*)/ ) {
```

Comment:
we have nicer \ readable way to do this formatting in perl?

Reply:
In that regex we have repeated patterns like "\s+(\d+\.\d+)", and would have liked to add something like (\s+(\d+\.\d+)){4} to represent \s+(\d+\.\d+)\s+(\d+\.\d+)\s+(\d+\.\d+)\s+(\d+\.\d+), but I was not able to make that work because we can't have nested ()'s and the ()'s save to a variable $1 or $2 and so on, depending on their position.  Maybe someone more well versed in regex can suggest something